qek/data/extractors.py

```diff
@@ -0,0 +1,432 @@
+        password: Your password on the Pasqal Cloud API. If you leave
+            this to None, you will need to enter your password manually.
+        device_name: The name of the device to use. As of this writing,
+            the default value of "FRESNEL" represents the latest QPU
+            available through the Pasqal Cloud API.
+        batch_id: Use this to resume a workflow e.g. after turning off
+            your computer while the QPU was executing your sequences.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        compiler: BaseGraphCompiler[GraphType],
+        project_id: str,
+        username: str,
+        password: str | None = None,
+        device_name: str = "FRESNEL",
+        batch_id: list[str] | None = None,
+    ):
+        sdk = SDK(username=username, project_id=project_id, password=password)
+
+        # Fetch the latest list of QPUs
+        specs = sdk.get_device_specs_dict()
+        device = cast(Device, deserialize_device(specs[device_name]))
+
+        super().__init__(path=path, device=device, compiler=compiler)
+        self._sdk = sdk
+        self._batch_id = batch_id
+
+    @property
+    def batch_ids(self) -> list[str] | None:
+        return self._batch_id
+
+    async def run(self) -> list[ProcessedData]:
+        if len(self.sequences) == 0:
+            logger.warning("No sequences to run, did you forget to call compile()?")
+            return []
+
+        batches: list[Batch] = []
+        if self._batch_id is None:
+            # Enqueue jobs.
+            self._batch_ids = []
+            for compiled in self.sequences:
+                logger.debug("Executing compiled graph #%s", id)
+                batch = self._sdk.create_batch(
+                    compiled.sequence.to_abstract_repr(),
+                    jobs=[{"runs": 1000}],
+                    wait=False,
+                )
+                logger.info(
+                    "Remote execution of compiled graph #%s starting, batched with id %s",
+                    id,
+                    batch.id,
+                )
+                batches.append(batch)
+                self._batch_ids.append(batch.id)
+            logger.info(
+                "All %s jobs enqueued for remote execution, with ids %s",
+                len(batches),
+                self._batch_ids,
+            )
+        else:
+            # Get jobs back from the cloud API.
+            for batch_id in self._batch_id:
+                batches.append(self._sdk.get_batch(batch_id))
+
+        # Now wait until all batches are complete.
+        waiting = True
+        while waiting:
+            waiting = False
+            for batch in batches:
+                if batch.status in {"PENDING", "RUNNING"}:
+                    # At least one job is pending, let's wait.
+                    await sleep(2)
+                    logger.debug("Job %s is still incomplete")
+                    waiting = True
```

Comment:
Unless I'm missing something, this is going to loop forever as we are not refreshing the batch data at each iteration

This may be a sign that we are missing a test for the qpu extractor; where we simulate the sequences execution taking a few iterations to run to completion

Reply:
Oh, yes, we're absolutely missing a test.
I'll try and find time to write one.